icicle_v3/backend/cpu/include/ntt_template.h

```diff
@@ -0,0 +1,264 @@
+#include "icicle/ntt.h"
+#include "icicle/errors.h"
+#include "icicle/runtime.h"
+#include "icicle/utils/log.h"
+
+#include "icicle/fields/field_config.h"
+#include "icicle/utils/log.h"
+
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <cmath>
+#include <cstdint>
+
+using namespace field_config;
+using namespace icicle;
+namespace ntt_template {
+  template <typename S>
+  class CpuNttDomain
+  {
+    // TODO - coset, mixed-radix NTT
+    int max_size = 0;
+    int max_log_size = 0;
+    S* twiddles = nullptr;
+
+  public:
+    template <typename U>
+    friend eIcicleError
+    cpu_ntt_init_domain(const Device& device, const U& primitive_root, const NTTInitDomainConfig& config);
+
+    template <typename U>
+    friend eIcicleError generate_twiddles(const U& primitive_root, U* twiddles, int logn);
+
+    template <typename U>
+    friend eIcicleError cpu_ntt_release_domain(const Device& device);
+
+    template <typename U, typename E>
+    friend eIcicleError
+    cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    template <typename U, typename E>
+    friend eIcicleError
+    cpu_ntt(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    S* get_twiddles() { return twiddles; }
+  };
+
+  template <typename S>
+  static CpuNttDomain<S>* s_ntt_domain = nullptr;
+
+  template <typename S = scalar_t>
+  eIcicleError generate_twiddles(const S& primitive_root, S* twiddles, int n)
+  {
+    S omega = primitive_root;
```

Comment:
(1) check that twiddles is not null and maybe that n is a power of two?
(2) in the CUDA implementation I think that they check more than that to make sure this really is a primitive root. I think it would be better to have it here too.